### DISCUSSION TITLE
dig vs fx - what the difference?

### DISCUSSION THREAD

**Discussion by maxgu** [uber-go/fx#951] [Q&A] [answered]
dig
> Bad for:
>    Exposing to user-land code as a Service Locator

Why?
fx and dig - have the same flow and case of use.

**Comment by sywhang** [comment, marked as answer]
Dig is a DI library. It has all the core logic Fx needs for actually performing DAG resolution of dependencies, type decorations and resolutions, etc. 

Fx is an DI-based application framework. Fx uses Dig to perform the core DI operations, and it comes with built-in utilities (such as Lifecycle and Loggers) that are useful when you are building an app. 

At Uber we use Fx to build services and libraries that powers most of our backend.
